india_compliance/gst_india/utils/test_e_invoice.py

```diff
@@ -0,0 +1,387 @@
+
+        self.assertFalse(
+            frappe.db.get_value(
+                "e-Waybill Log", {"reference_name": debit_note.name}, "name"
+            )
+        )
+
+    @responses.activate
+    def test_cancel_e_invoice(self):
+        """Test for generate and cancel e-Invoice
+        - Test function `validate_if_e_invoice_can_be_cancelled`
+        """
+
+        test_data = self.e_invoice_test_data.get("goods_item_with_ewaybill")
+        si = create_sales_invoice(**test_data.get("kwargs"))
+
+        self.assertRaisesRegex(
+            frappe.exceptions.ValidationError,
+            re.compile(r"^(IRN not found)$"),
+            validate_if_e_invoice_can_be_cancelled,
+            si,
+        )
+
+        # Mock response for generating irn
+        result = self._mock_e_invoice_response(si, test_data)
+
+        self._generate_e_invoice(si, result)
+        log_and_process_e_waybill_generation(si, result, with_irn=True)
+
+        si_doc = load_doc("Sales Invoice", si.name, "cancel")
+
+        si_doc.get_onload().get("e_invoice_info", {}).update({"acknowledged_on": None})
+
+        self.assertRaisesRegex(
+            frappe.exceptions.ValidationError,
+            re.compile(r"^(e-Invoice can only be cancelled.*)$"),
+            validate_if_e_invoice_can_be_cancelled,
+            si_doc,
+        )
+
+        values = frappe._dict(
+            {"reason": "Data Entry Mistake", "remark": "Data Entry Mistake"}
+        )
+
+        self._cancel_e_waybill(si_doc, values)
+        self._cancel_e_invoice(si_doc, values)
+
+        self.assertDocumentEqual({"einvoice_status": "Cancelled", "irn": ""}, si_doc)
+        self.assertDocumentEqual({"ewaybill": ""}, si_doc)
+
+    def test_validate_e_invoice_applicability(self):
+        """Test if e_invoicing is applicable"""
+
+        si = create_sales_invoice(
+            customer="_Test Unregistered Customer",
+            gst_category="Unregistered",
+            do_not_submit=True,
+        )
+        self.assertRaisesRegex(
+            frappe.exceptions.ValidationError,
+            re.compile(r"^(e-Invoice is not applicable for invoices.*)$"),
+            validate_e_invoice_applicability,
+            si,
+        )
+
+        si.update(
+            {
+                "gst_category": "Registered Regular",
+                "customer": "_Test Registered Customer",
+            }
+        )
+        si.save(ignore_permissions=True)
+        frappe.db.set_single_value(
+            "GST Settings", "e_invoice_applicable_from", "2045-05-18"
+        )
+
+        self.assertRaisesRegex(
+            frappe.exceptions.ValidationError,
+            re.compile(r"^(e-Invoice is not applicable for invoices before.*)$"),
+            validate_e_invoice_applicability,
+            si,
+        )
+
+        frappe.db.set_single_value(
+            "GST Settings", "e_invoice_applicable_from", get_datetime()
+        )
+
+    def _generate_e_invoice(self, doc, result):
```

Comment:
Why are you duplicating these functions in test cases?

What is the objective of a test case if you are testing a test function?

Reply:
Yes, I just observe it yesterday. I will update it.